File: leetcode/C++/palindrome-pairs.cpp
class TrieNode {
public:
    int idx;
    unordered_set<int> us;
    unordered_map<char,TrieNode*> um;
    TrieNode() {idx=-1;}
};

class Trie {
public:
    TrieNode *node;
    Trie() {node=new TrieNode();}
    bool palindrome(string &w,int &s, int &l){
        for(int i=0; i<(l-s)/2; ++i){
            if(w[s+i]!=w[l-1-i]) return false;
        }
        return true;
    }
    void insert(int &idx, string &w){
        TrieNode *cur = node;
        int s=0;
        for(int i=w.length(); i>0; --i){
            if(palindrome(w,s, i)) {
                cur->us.insert(idx);
            }
            if(cur->um.find(w[i-1])==cur->um.end()){
                cur->um[w[i-1]]=new TrieNode();
            }
            cur=cur->um[w[i-1]];
        }
        cur->idx=idx;
    }
    void search(int &idx, string &w, vector<vector<int>> &res){
        TrieNode *cur = node;
        int sz = w.length();
        for(int i=0; i<w.length(); ++i){
            if(cur->idx >= 0){
                if(palindrome(w,i,sz)){
                    res.push_back({idx, cur->idx});
                }
            }
            if(cur->um.find(w[i])==cur->um.end()) return;
            cur=cur->um[w[i]];
        }
        if(cur->idx >= 0 && cur->idx != idx){
            res.push_back({idx, cur->idx});
        }
        for(const int &t : cur->us){
            res.push_back({idx,t});
        }
    }
};

class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        vector<vector<int>> sol;
        Trie trie = Trie();
        for(int i=0; i<words.size(); ++i) {
            trie.insert(i,words[i]);
        }
        for(int i=0; i<words.size(); ++i){
            trie.search(i, words[i], sol);
        }
        return sol;
    }
};

File: leetcode/Python/palindrome-pairs.py
class TrieNode:
    def __init__(self):
        self.idx = -1
        self.us = []
        self.um = dict()
class Trie:
    def __init__(self):
        self.node = TrieNode()
    def palindrome(self, w):
        return w[::-1]==w
    def insert(self, idx, w):
        cur = self.node
        for i in range(len(w),0,-1):
            if self.palindrome(w[:i]):
                cur.us.append(idx)
            if w[i-1] not in cur.um:
                cur.um[w[i-1]]=TrieNode()
            cur=cur.um[w[i-1]]
        cur.idx=idx
    def search(self, idx, w, res):
        cur=self.node
        for i in range(len(w)):
            if cur.idx>=0:
                if self.palindrome(w[i:]):
                    res.append([idx, cur.idx])
            if w[i] not in cur.um:
                return
            cur=cur.um[w[i]]
        if cur.idx >= 0 and cur.idx != idx:
            res.append([idx, cur.idx])
        for t in cur.us:
            if t == idx:
                 continue
            res.append([idx, t])
class Solution:
    def palindromePairs(self, words: List[str]) -> List[List[int]]:
        sol ,t = [], Trie()
        for i, w in enumerate(words):
            t.insert(i,w)
        for i, w in enumerate(words):
            t.search(i,w,sol)
        return sol
        